### DISCUSSION TITLE
Contact and Contact Group suggestions and musings

### DISCUSSION THREAD

**Discussion by ed-ud** [netbox-community/netbox#15064] [Ideas]
I find myself wanting to use the Contacts features of NetBox for more than just a static list of organizational people to assign to various objects as contacts.  To that end, there are a few features that would be nice to have:

1. Allow a Contact to be in more than one Contact Group.  We have organizational (HR) departments which are useful as the "official word", but we also have teams/groups which sometimes involve people from more than one organizational department such that we can't just use nesting to have a person appear to be in the groups we want.  It's handy to be able to have groups that encompass any set of people we want to include, regardless of what organizational department they happen to be in.

2. Allow Contact Groups to be used interchangeably with Contacts in Contact Assignments.  This would be similar to how Contact Assignments can already be applied to many different object types.  We usually have groups that are responsible for things, rather than specific people, and the membership in our groups can change.  The groups we're using for actual assignments aren't typically our "HR" departments and involve a few people or a dozen - for example, the same 10 people are generally responsible for many things but when that list of people changes or priorities shift, we have to add/remove a lot of assignments for all of those things, instead of just updating the Contact Group membership and having maybe a person or two listed as Primary.

3. Allow Services to have Contact Assignments.  In the absence of an "Application" or "Application Cluster" model to encompass, e.g., a bunch of software application instances running on different machines for a PostgreSQL database cluster or a set of Devices/VMs/Services that encompass an Apache web server farm, etc., and document who is responsible for the overall "application", I find myself wanting to assign contacts to the individual Services to indicate who runs the software.

On a related note, I've treated Tenants as "the person using something" for external vendors, guests, etc., of people who aren't technically affiliated with our organization.  But for certain things like IP addresses, we "allocate" those to Tenants, but also to organizational people defined as Contacts.  I'm not sure if there is a better way than just putting a Contact Assignment on the IP that has a role like "Allocated to User" or such to mark an IP address as being assigned to a person who is part of the organization.  There have been discussions about marking a Prefix with an indicator for who manages the IPs in the Prefix, and we are facing the same issue - we can certainly assign a contact with a role "Prefix Manager" for Prefixes where we might need to decide whether a given person is allowed to allocate an IP in that Prefix (not necessarily using NetBox directly, likely via a ticket request and a program that needs to decide if it is allowed, allocate the next available IP, mark who is going to be using that IP, register a DNS name (?), post it back to the ticket... some process like that, which is currently not very well defined for us yet), but is there a better way?

**Comment by ed-ud** [comment]
As a workaround, I've added a custom field "aux_groups" to Contacts that is a list of Contact Group objects, and a program can query NetBox to fetch a Contact Group and determine who are listed as members of that group.  The drawback is that the Contact Group, when viewed in NetBox, doesn't show any Contacts under Related Objects (because there aren't any), so I added a Custom Link to Contact Groups that will look up the Contacts that have the group in their list.  So not completely awful, but still pretty hacky.

**Comment by ed-ud** [comment]
Is there a way to add to the "Related Objects" list on a given view.  For example, if I refer to a Contact Group in a custom field, can I make the Contact Group view show objects that refer to the Contact Group besides just the Contacts that refer to the Contact Group?

**Comment by ed-ud** [comment]
Anyone else like any of the OP suggestions?

Regarding the "Contact and Contact Group used interchangeably in Contact Assignments", an alternative would be to have a new model "Contact Group Assignments" that use a Contact Group and Contact Role instead of a single Contact.

**Reply by mcdonalm-at-frib** [reply]
This is an old(er) discussion, but I definitely am interesting in the idea of being able to use contact groups as an assignment. Our team of network engineers all equally respond / administer our equipment, but like you: the staff list of who is in that group changes.

**Reply by sol1-matt** [reply]
> This is an old(er) discussion, but I definitely am interesting in the idea of being able to use contact groups as an assignment. Our team of network engineers all equally respond / administer our equipment, but like you: the staff list of who is in that group changes.

The problem with contact groups is the relationship from the contact to the contact group is one to one
this means a contact can only belong to a single group.

A distribution list is the other direction, the relationship is from distribution list to multiple contacts. This allows for the same contact be belong to multiple contact groups. 

ed-ud's [workaround](https://github.com/netbox-community/netbox/discussions/15064#discussioncomment-8388734) works well.

You create single custom field on Contact of type `Multiple Objects (Contact)`, I called mine `Distribution List Contacts`.
I also prefix the Contacts that are distribution lists with Distribution List  so the name is `Distribution List: <listname>`.

**Reply by jeremystretch** [reply]
> The problem with contact groups is the relationship from the contact to the contact group is one to one
> this means a contact can only belong to a single group.

FYI this was just changed in NetBox v4.3.0. A contact can now be assigned to multiple groups (see #17170).

**Reply by ed-ud** [reply]
That's GREAT... thanks!

Now if we could do the same for VLANs being a member of more than one VLANGroup... :-D

**Reply by ed-ud** [reply]
Maybe the easiest solution for assigning groups to things is to have another model Contact Group Assignments that works the same way as the existing Contact Assignments, although I would guess that making a Contact Group field in the existing Contact Assignments model, with some added logic to assign either a Contact or Contact Group depending on how you fill out the Add Contact form (which already has a Contact Group field that is used as a filter to limit which Contacts are shown in the Contact field) would work as well (and avoid having to create a new model, more tables to show the two types of assignments on the Contacts tab, etc.)

**Reply by sol1-matt** [reply]
@ed-ud you mean the `contact group` is the assignment not the `contact`?

That would work, another option would be a `distribution list` object which works like the contact object but has a list of contacts on it (like the custom field solution you suggested). 

I think the `distribution list` solution has the benefit of being explicit. eg: "This object is used to link to objects for the purpose of assigning x contacts to said object and that is it's only job". 

Where as a `contact group assignment` means the `contact group` has additional functionality above it's current use case.
